### PR TITLE
Add KB: Troubleshooting: NFS server setting and mount a NFS storage

### PR DESCRIPTION
**Pull request by mantissahz**
[longhorn/longhorn#3576](https://github.com/longhorn/longhorn/issues/3576)

### CODE REVIEW

**Comment by derekbit** [comment]
@mantissahz 
You can also add the `permission denied` into the doc
ref. https://github.com/longhorn/longhorn/issues/3416

File: content/kb/troubleshooting-nfs-server-setting-and-mount-nfs.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+### Reason
+
+The NFS client has no permission to access remote directory the NFS server shared. The directory is not accessiable by normal group/user and the option "root_squash" might be used for shared the directory.
+### Solution
+
+1. Use the option "no_root_squash" instead of "root_squash"
+
+2. Add chmod o+w [shared directory path] or change the directory owner to `nobody`.
```
Comment (innobead) [inline content/kb/troubleshooting-nfs-server-setting-and-mount-nfs.md:82]:
```suggestion
2. Execute `chmod o+w [exported directory path]` or change the owner of the directory to `nobody`.
```

File: content/kb/troubleshooting-nfs-server-setting-and-mount-nfs.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+## Related information
+
+* Related Longhorn issue:</br>
```
Comment (innobead) [inline content/kb/troubleshooting-nfs-server-setting-and-mount-nfs.md:86]:
```suggestion
* Related Longhorn issues:
```

File: content/kb/troubleshooting-nfs-server-setting-and-mount-nfs.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  <https://github.com/longhorn/longhorn/issues/3576>
+  <https://github.com/longhorn/longhorn/discussions/3805>
```
Comment (innobead) [inline content/kb/troubleshooting-nfs-server-setting-and-mount-nfs.md:88]:
```suggestion

  - <https://github.com/longhorn/longhorn/issues/3576>
  - <https://github.com/longhorn/longhorn/discussions/3805>
```

**Review by innobead — CHANGES_REQUESTED** [review]

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  - "backup"
+---
+
+## Applicable versions
+
+All Longhorn versions.
+
+## Prerequisite
+
+Longhorn only supports NFS v4.0, v4.1, v4.2.
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:16]:
```suggestion
Longhorn only supports NFS v4.0, v4.1 and v4.2.
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
+---
+title: "Troubleshooting: Unable to mount an NFS backup target"
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:2]:
```suggestion
title: "Troubleshooting: Unable to access an NFS backup target"
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+```bash
+    #/proc/fs/nfsd/versions
+```
+
+* Client side:</br>
+What versions the NFS mount is configured to support:
+
+```bash
+    #nfsstat -m
+```
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]:
```suggestion
* Server side:

The NFS server can support multiple versions by the configuration, so using which version for each connection will actually depend on how the NFS share is mounted by the client .</br>
What versions the server is configured to support:

```bash
    #/proc/fs/nfsd/versions
```

* Client side:
* 
What versions the NFS mount is configured to support:

```bash
    #nfsstat -m
```
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+## Symptom 1
+
+After setting up an NFS server, `/etc/exports` is as below:
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:37]:
```suggestion
Setting up an NFS server and `/etc/exports` is as the below example:
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+### How to determine which NFS version is used
+
+* Server side:</br>
+The NFS server can support multiple versions by the configuration, so using which version for each connection will actually depend on how the NFS share is mounted by the client .</br>
+What versions the server is configured to support:
+
+```bash
+    #/proc/fs/nfsd/versions
+```
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:26]:
```suggestion
    ```bash
    # /proc/fs/nfsd/versions
    ```
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+### Reason
+
+For NFSv4, if the **fsid=0** or **fsid=root** option is configured with the exported directory in `/etc/exports`, it means the exported directory is the root of all exported filesystems. The example here is `/opt/nfs-server`.
+
+If users try to use the absolute path of the exported directory to list backups, that mounting error will happen.
+
+### Solution
+
+1. Mount with the absolute path of the exported directory (ex: `/opt/nfs-server`) on the client side **without** the option "fsid=0" or "fsid=root" on the server side.
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:64]:
```suggestion
1. Mount with the absolute path of the exported directory (ex: `/opt/nfs-server`) on the client side **without** the option `fsid=0` or `fsid=root` on the server side.
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 6 above)
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:33]: Same as above

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+2. Mount with the path "/" on the client side still **with** the option "fsid=0" or "fsid=root" on the server side.
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:66]:
```suggestion
2. Mount with the path "/" on the client side still **with** the option `fsid=0` or `fsid=root` on the server side.
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+There will be an error about mounting failure.
+
+```text
+, error exit status 32: vers=4.2: Failed to execute: mount [-t nfs4 -o nfsvers=4.2 -o actimeo=1 192.168.121.170:/opt/nfs-server /var/lib/longhorn-backupstore-mounts/192_168_121_170/opt/nfs-server], output mount.nfs4: mounting 192.168.121.170:/opt/nfs-server failed, reason given by server: No such file or directory
+, error exit status 32: Cannot mount using NFSv4
+```
+
+### Reason
+
+For NFSv4, if the **fsid=0** or **fsid=root** option is configured with the exported directory in `/etc/exports`, it means the exported directory is the root of all exported filesystems. The example here is `/opt/nfs-server`.
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:58]:
```suggestion
For NFSv4, if the `fsid=0` or `fsid=root` option is configured with the exported directory in `/etc/exports`, it means the exported directory is the root of all exported filesystems. The example here is `/opt/nfs-server`.
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+After updating an NFS target endpoint as the backup target, then start backup operations.
+
+```text
+error running create backup command: failed to create backup to nfs://192.168.121.170:/opt/nfs-server for volume test-for-backup: rpc error: code = Unknown desc = mkdir /var/lib/backupstore/192_168_121_170/opt/nfs-server/backup/longhorn/backupstore: permission denied" , error exit status 1
+```
+
+### Reason
+
+The directory is not accessible by a non-root user due to the option "root_squash" might be used for the exported directory.
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:78]:
```suggestion
The directory is not accessible by a non-root user due to the option `root_squash` might be used for the exported directory.
```

**Review by innobead — CHANGES_REQUESTED** [review]

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+```bash
+    # /proc/fs/nfsd/versions
+```
+
+* Client side:</br>
+What versions the NFS mount is configured to support:
+
+```bash
+    # nfsstat -m
+```
```
Comment (innobead) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:33]:
```suggestion
* Server side:

The NFS server can support multiple versions by the configuration, so using which version for each connection will actually depend on how the NFS share is mounted by the client .</br>
What versions the server is configured to support:

```bash
    # /proc/fs/nfsd/versions
```

* Client side:

What versions the NFS mount is configured to support:

```bash
    # nfsstat -m
```
```

**Review by innobead — CHANGES_REQUESTED** [review]

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 6 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: Modified, thanks

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 8 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: Modified, thanks

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 13 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: Modified, thanks

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+* Server side:
+
+The NFS server can support multiple versions by the configuration, so using which version for each connection will actually depend on how the NFS share is mounted by the client.
+
+What versions the server is configured to support:
+
+```bash
+  # cat /proc/fs/nfsd/versions
+```
+
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:29]: Simply mentioning the purpose of the command is enough.

```suggestion
Check which versions of NFS are currently enabled.

```bash
  # /proc/fs/nfsd/versions
```

```

**Review by derekbit — CHANGES_REQUESTED** [review]

**Review by innobead — DISMISSED** [review]

**Comment by innobead** [comment]
@derekbit has other comments, please fix them first before merging. Thanks.

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+## Applicable versions
+
+All Longhorn versions.
+
+## Prerequisite
+
+Longhorn only supports NFS v4.0, v4.1 and v4.2.
+
+### How to determine which NFS version is used
+
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:19]: in Prerequisite? or this section

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 14 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: @derekbit could you take a look again please?

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  - "backup"
+---
+
+## Applicable versions
+
+All Longhorn versions.
+
+## Prerequisite
+
+Longhorn only supports NFS v4.0, v4.1 and v4.2.
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:16]:
```suggestion
Longhorn supports only NFS versions 4.0, 4.1, and 4.2. Users should ensure that both their backup server and client support one of the supported versions.
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+### How to determine which NFS version is used
+
+Longhorn supports only NFS versions 4.0, 4.1, and 4.2. Users should ensure that both their backup server and client support one of the supported versions.
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:20]: Redundant.

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+* Server side:
+
+The NFS server can support multiple versions by the configuration, so using which version for each connection will actually depend on how the NFS share is mounted by the client.
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:24]: Remove this part. Just mentioning how to check is enough.

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 16 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: Modified, thanks

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 17 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: Deleted, thanks

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 18 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md]: ok, deleted

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+* Client side:
+
+What versions the NFS mount is configured to support:
+
+```bash
+  # nfsstat -m
+```
+
+## Symptom 1
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:36]: We can describe the symptom and then provide possible root causes and solutions.

For example,
- Symptom
  The error message on backup UI page complained
  ```
  error exit status 32: vers=4.2: Failed to execute: mount [-t nfs4 -o nfsvers=4.2 -o actimeo=1 192.168.121.170:/opt/nfs-server /var/lib/longhorn-backupstore-mounts/192_168_121_170/opt/nfs-server], output mount.nfs4: mounting 192.168.121.170:/opt/nfs-server failed, reason given by server: No such file or directory
  , error exit status 32: Cannot mount using NFSv4"
  ```

- Troubleshooting
  - Possible Reason 1
    Check if the configuration in `/etc/exports` contains `fsid=0`....
    If YES, explain why and how to resolve it.
  - Possible Reason 2
     ....

**Review by derekbit — CHANGES_REQUESTED** [review]

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+If users try to use the absolute path of the exported directory to list backups, that mounting error will happen.
+
+### Solution
+
+1. Mount with the absolute path of the exported directory (ex: `/opt/nfs-server`) on the client side **without** the option `fsid=0` or `fsid=root` on the server side.
+
+2. Mount with the path "/" on the client side still **with** the option `fsid=0` or `fsid=root` on the server side.
+
+## Symptom 2
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:69]: This is an good example for rewriting Symptom 1.

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+After updating an NFS target endpoint as the backup target, then start backup operations.
+
+```text
+error running create backup command: failed to create backup to nfs://192.168.121.170:/opt/nfs-server for volume test-for-backup: rpc error: code = Unknown desc = mkdir /var/lib/backupstore/192_168_121_170/opt/nfs-server/backup/longhorn/backupstore: permission denied" , error exit status 1
+```
+
+### Reason
+
+The directory is not accessible by a non-root user due to the option `root_squash` might be used for the exported directory.
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:79]:
```suggestion
The exported directory is not accessible by non-root users due to the `root_squash` option being used.
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+
+### Solution
+
+1. Mount with the absolute path of the exported directory (ex: `/opt/nfs-server`) on the client side **without** the option `fsid=0` or `fsid=root` on the server side.
+
+2. Mount with the path "/" on the client side still **with** the option `fsid=0` or `fsid=root` on the server side.
+
+## Symptom 2
+
+After updating an NFS target endpoint as the backup target, then start backup operations.
```
Comment (derekbit) [inline content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:71]: Simply mention the symptom is fine.
```suggestion
The Backup page pops up with a permission denied error, for example,
```

File: content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md (hunk 19 above)
Comment (mantissahz) [reply content/kb/troubleshooting-unable-to-mount-an-nfs-backup-target.md:36]: @derekbit I rewrote Symptom 1 and does it look like what you mentioned?

**Review by derekbit — APPROVED** [review]
In general, LGTM.

**Review by innobead — APPROVED** [review]
LGTM.

**Pull request merged**